pkg/phlaredb/schemas/v1/profiles.go

```diff
@@ -28,6 +28,7 @@ var (
 		phlareparquet.NewGroupField("StacktraceID", parquet.Encoded(parquet.Uint(64), &parquet.DeltaBinaryPacked)),
 		phlareparquet.NewGroupField("Value", parquet.Encoded(parquet.Int(64), &parquet.DeltaBinaryPacked)),
 		phlareparquet.NewGroupField("Labels", pprofLabels),
+		phlareparquet.NewGroupField("SpanID", parquet.Optional(parquet.Encoded(parquet.Uint(64), &parquet.Plain))),
```

Comment:
I wonder if a dictionary encoding makes sense here, I would expect many rows to have similar columns

Reply:
Last time I checked statistics we have, a span had 1.5 stack trace samples on average :) There are many factors, we should experiment with the real data. Ideally, the encoding should be adaptive.

The [benchmark](https://github.com/grafana/pyroscope/blob/6eba1494e7ca669ed098726d3c40e18c798a8538/pkg/phlaredb/schemas/v1/profiles_test.go#L311) generates 1K profiles with 100 samples each. Some fraction of the samples have unique span IDs, the rest use duplicate values (encoded in the benchmark name: `{ratio}_({unique}/{total})`).

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/pyroscope/pkg/phlaredb/schemas/v1
                                  │  plain.txt   │            rle-dict.txt             │
                                  │    sec/op    │   sec/op     vs base                │
_SpanID_Encoding/1_(100/100)-10     3.987m ±  1%   4.414m ± 4%  +10.69% (p=0.000 n=10)
_SpanID_Encoding/0.5_(50/100)-10    3.988m ±  1%   4.204m ± 3%   +5.40% (p=0.000 n=10)
_SpanID_Encoding/0.25_(25/100)-10   3.984m ±  0%   4.081m ± 4%   +2.43% (p=0.001 n=10)
_SpanID_Encoding/0.15_(15/100)-10   4.008m ±  2%   4.012m ± 2%        ~ (p=1.000 n=10)
_SpanID_Encoding/0.05_(5/100)-10    4.103m ± 12%   4.064m ± 1%        ~ (p=0.631 n=10)
geomean                             4.014m         4.153m        +3.46%

                                  │   plain.txt   │             rle-dict.txt             │
                                  │     bytes     │    bytes      vs base                │
_SpanID_Encoding/1_(100/100)-10      1.057Mi ± 0%   1.254Mi ± 0%  +18.65% (p=0.000 n=10)
_SpanID_Encoding/0.5_(50/100)-10    1081.8Ki ± 0%   880.7Ki ± 0%  -18.59% (p=0.000 n=10)
_SpanID_Encoding/0.25_(25/100)-10   1082.2Ki ± 0%   673.6Ki ± 0%  -37.75% (p=0.000 n=10)
_SpanID_Encoding/0.15_(15/100)-10   1082.4Ki ± 0%   583.6Ki ± 0%  -46.09% (p=0.000 n=10)
_SpanID_Encoding/0.05_(5/100)-10    1082.4Ki ± 0%   387.9Ki ± 0%  -64.17% (p=0.000 n=10)
geomean                              1.057Mi        703.6Ki       -34.99%

                                  │  plain.txt   │             rle-dict.txt             │
                                  │     B/op     │     B/op      vs base                │
_SpanID_Encoding/1_(100/100)-10     1.818Mi ± 0%   2.683Mi ± 1%  +47.60% (p=0.000 n=10)
_SpanID_Encoding/0.5_(50/100)-10    1.819Mi ± 0%   2.279Mi ± 1%  +25.30% (p=0.000 n=10)
_SpanID_Encoding/0.25_(25/100)-10   1.818Mi ± 0%   2.035Mi ± 0%  +11.90% (p=0.000 n=10)
_SpanID_Encoding/0.15_(15/100)-10   1.819Mi ± 0%   1.951Mi ± 0%   +7.26% (p=0.000 n=10)
_SpanID_Encoding/0.05_(5/100)-10    1.819Mi ± 0%   1.859Mi ± 0%   +2.21% (p=0.000 n=10)
geomean                             1.818Mi        2.142Mi       +17.80%

                                  │  plain.txt  │            rle-dict.txt            │
                                  │  allocs/op  │  allocs/op   vs base               │
_SpanID_Encoding/1_(100/100)-10     1.744k ± 0%   1.742k ± 0%  -0.11% (p=0.000 n=10)
_SpanID_Encoding/0.5_(50/100)-10    1.744k ± 0%   1.745k ± 0%  +0.06% (p=0.008 n=10)
_SpanID_Encoding/0.25_(25/100)-10   1.744k ± 0%   1.745k ± 0%  +0.06% (p=0.000 n=10)
_SpanID_Encoding/0.15_(15/100)-10   1.744k ± 0%   1.746k ± 0%  +0.11% (p=0.000 n=10)
_SpanID_Encoding/0.05_(5/100)-10    1.744k ± 0%   1.746k ± 0%  +0.11% (p=0.000 n=10)
geomean                             1.744k        1.745k       +0.05%
```

Reply:
Experimented with some more data and I can say that use of dictionaries is fully justified. Later, when/if we will record all spans (not just the root), we can revisit this decision:

```
        Row group: 2
                 Row Count: 99627
                 Row size: 101 MB
                 Columns:
+--------------------------------------------------+----------------------+----------+---------------------+-----------------------+-------------+-------+-----------+-------------+
|                       COL                        |         TYPE         |  NUMVAL  | TOTALCOMPRESSEDSIZE | TOTALUNCOMPRESSEDSIZE | COMPRESSION |   %   | PAGECOUNT | AVGPAGESIZE |
+--------------------------------------------------+----------------------+----------+---------------------+-----------------------+-------------+-------+-----------+-------------+
| ID                                               | FIXED_LEN_BYTE_ARRAY |    99627 | 1.6 MB              | 1.6 MB                |        0.00 |  1.58 |         1 | 1.6 MB      |
| SeriesIndex                                      | INT32                |    99627 | 6.1 kB              | 6.1 kB                |        0.00 |  0.01 |         1 | 8.8 kB      |
| StacktracePartition                              | INT64                |    99627 | 20 kB               | 20 kB                 |        0.00 |  0.02 |         1 | 68 kB       |
| TotalValue                                       | INT64                |    99627 | 159 kB              | 159 kB                |        0.00 |  0.16 |         1 | 314 kB      |
| Samples/list/element/StacktraceID                | INT64                | 22186975 | 33 MB               | 33 MB                 |        0.00 | 32.33 |        36 | 401 kB      |
| Samples/list/element/Value                       | INT64                | 22186975 | 26 MB               | 26 MB                 |        0.00 | 25.77 |        36 | 653 kB      |
| Samples/list/element/Labels/list/element/Key     | INT64                | 22186975 | 485 kB              | 485 kB                |        0.00 |  0.48 |         8 | 67 kB       |
| Samples/list/element/Labels/list/element/Str     | INT64                | 22186975 | 485 kB              | 485 kB                |        0.00 |  0.48 |         8 | 67 kB       |
| Samples/list/element/Labels/list/element/Num     | INT64                | 22186975 | 485 kB              | 485 kB                |        0.00 |  0.48 |         8 | 67 kB       |
| Samples/list/element/Labels/list/element/NumUnit | INT64                | 22186975 | 485 kB              | 485 kB                |        0.00 |  0.48 |         8 | 67 kB       |
| Samples/list/element/SpanID                      | INT64                | 22186975 | 37 MB               | 37 MB                 |        0.00 | 36.63 |        14 | 724 kB      |
| DropFrames                                       | INT64                |    99627 | 45 B                | 45 B                  |        0.00 |  0.00 |         1 | 43 B        |
| KeepFrames                                       | INT64                |    99627 | 45 B                | 45 B                  |        0.00 |  0.00 |         1 | 43 B        |
| TimeNanos                                        | INT64                |    99627 | 797 kB              | 797 kB                |        0.00 |  0.79 |         1 | 800 kB      |
| DurationNanos                                    | INT64                |    99627 | 797 kB              | 797 kB                |        0.00 |  0.79 |         1 | 38 B        |
| Period                                           | INT64                |    99627 | 40 B                | 40 B                  |        0.00 |  0.00 |         1 | 38 B        |
| Comments/list/element                            | INT64                |    99627 | 50 B                | 50 B                  |        0.00 |  0.00 |         1 | 47 B        |
| DefaultSampleType                                | INT64                |    99627 | 40 B                | 40 B                  |        0.00 |  0.00 |         1 | 38 B        |
+--------------------------------------------------+----------------------+----------+---------------------+-----------------------+-------------+-------+-----------+-------------+
```